ui/dashboards/src/components/Variables/VariableEditorForm.tsx

```diff
@@ -0,0 +1,306 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  Box,
+  Typography,
+  Switch,
+  TextField,
+  Grid,
+  FormControl,
+  InputLabel,
+  MenuItem,
+  Select,
+  Button,
+  Stack,
+  Tooltip,
+} from '@mui/material';
+import { useImmer } from 'use-immer';
+import { useListPluginMetadata, PluginSpecEditor, usePluginRegistry } from '@perses-dev/plugin-system';
+import { VariableDefinition, TextVariableDefinition, ListVariableDefinition } from '@perses-dev/core';
+import React from 'react';
+
+const VARIABLE_TYPES = ['ListVariable', 'TextVariable'] as const;
+
+function getInitialState(initialVariableDefinition: VariableDefinition) {
+  const textVariableFields = {
+    value: (initialVariableDefinition as TextVariableDefinition).spec.value ?? '',
+  };
+
+  const listVariableFields = {
+    allowMultiple: false,
+    allowAll: false,
+    plugin: {
+      kind: '',
+      spec: {},
+    },
+  };
+  if (initialVariableDefinition.kind === 'ListVariable') {
+    listVariableFields.allowMultiple = initialVariableDefinition.spec.allow_all_value ?? false;
+    listVariableFields.allowAll = initialVariableDefinition.spec.allow_all_value ?? false;
+    listVariableFields.plugin = initialVariableDefinition.spec.plugin;
+  }
+
+  return {
+    name: initialVariableDefinition.spec.name,
+    label: initialVariableDefinition.spec.display?.label,
+    kind: initialVariableDefinition.kind,
+    description: '',
+    listVariableFields,
+    textVariableFields,
+  };
+}
+
+type VariableEditorState = ReturnType<typeof getInitialState>;
+
+function getVariableDefinitionFromState(state: VariableEditorState): VariableDefinition {
+  const { name, label, kind } = state;
+
+  const commonSpec = {
+    name,
+    display: {
+      label,
+    },
+  };
+
+  if (kind === 'TextVariable') {
+    return {
+      kind,
+      spec: {
+        ...commonSpec,
+        ...state.textVariableFields,
+      },
+    } as TextVariableDefinition;
```

Comment:
Do you actually need the casts in here? Just always makes me a little leery and try to avoid if possible since it can hide type errors. One possible alternative (more verbose, but avoids the cast) is to use a temp variable with a type like:

```ts
const textVariable: TextVariableDefinition = {
  // ... insert construction of value here
};
return textVariable;
```